hpcbench/campaign.py

```diff
@@ -225,18 +225,60 @@ def _expand_tag_pattern(cls, pattern):
             if mode == 'match':
                 pattern[mode] = re.compile(pattern[mode])
             elif mode == 'nodes':
-                pattern[mode] = NetworkConfig._expand_nodes(pattern[mode])
+                pattern[mode] = cls._expand_nodes(pattern[mode])
+            elif mode == 'tags':
+                pass  # don't fail but ignore tags
             else:
                 raise Exception('Unknown tag association pattern: %s',
                                 mode)
 
-    def _expand_tag(self, tag):
-        config = self.network.tags[tag]
-        if isinstance(config, dict):
-            config = [config]
-            self.network.tags[tag] = config
-        for pattern in config:
-            self._expand_tag_pattern(pattern)
+    @classmethod
+    def _is_leaf(cls, config):
+        # returns True if in none of the modes and patterns is 'tags'
+        return reduce(lambda x, y: x & y,
+                      map(lambda p: 'tags' not in p.keys(), config))
+
+    @classmethod
+    def _resolve(cls, tag, config, expanded, recursive, visited):
+        for pattern in config[:]:
+            # we work with a copy so we can modify the original
+            # first expand all the other modes
+            cls._expand_tag_pattern(pattern)
+            # now let's go through that tags if they exist in this pattern
+            if 'tags' in list(pattern):
+                tags = pattern['tags']
+                for rectag in tags:
+                    if rectag in expanded:
+                        config += expanded[rectag]
+                    elif rectag in visited:
+                        raise Exception('found circular dependency')
+                    else:
+                        recconfig = recursive.pop(rectag)
+                        visited.add(rectag)
+                        cls._resolve(rectag, recconfig,
+                                     expanded, recursive, visited)
+                pattern.pop('tags')  # we've expanded this, it can be deleted
+        expanded[tag] = config
+
+    def _expand_tags(self):
+        expanded = {}
+        recursive = {}
+        for tag, config in self.network.tags.items():
+            if isinstance(config, dict):
+                config = [config]
+            if NetworkConfig._is_leaf(config):
+                for pattern in config:
+                    NetworkConfig._expand_tag_pattern(pattern)
+                expanded[tag] = config
+            else:
+                recursive[tag] = config
+        # we finished all the leafs (tags without any recursive tag references)
+        visited = set(expanded.keys())
+        while len(recursive) > 0:
```

Comment:
empty dictionary evaluates to `False`. You can use `while recursive:`

Reply:
Oh of course! I meant to change this :)